### ISSUE TITLE
qx seems to loose the subpr. way for print /dev/tty

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8374]
Migrated from [rt.perl.org#38748](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=38748) (status was 'open')

Searchable as RT38748$

**Comment by p5pRT** [comment]
### From ende@telefonica.net

#### Created by juanfc@lcc.uma.es



I send messages through /dev/tty with

&nbsp;    exec echo >/dev/tty

in a tcl script that is called from a Perl script\.

The Perl script is on forever with a loop\, allowing me
a more easy control of both the Tcl script and the frequency
and actions I want the daemon do each 10 minutes\.



In the "daemonized" perl script\, that has the loop&#8203;:

while \(1\) \{
&nbsp;     \`/Users/home/bin/tclprocThatusesdevtty\.tcl \-a\`;
&nbsp;     \# \.\.\.\.\.
&nbsp;     sleep 600;
\}

only the first time the Tcl program is runned\, the messages sent
by the Tcl program are really sent to the /dev/tty\.  Then\, no more
messages are seen in the Terminal\.  The perl script seems
running ok all the time and the Tcl program also seems been called\.\.\.

If I do the same stuff with sh

while \[ 1 \-eq 1 \]; do
&nbsp;     /Users/juanfc/bin/drg\.tcl \-a
&nbsp;     \# \.\.\.\.
&nbsp;     sleep 600
done


there is no problem at all\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=high

Site configuration information for perl v5.8.6:

Configured by root at Sun Mar 20 16:31:42 PST 2005.

Summary of my perl5 (revision 5 version 8 subversion 6) configuration:
   Platform:
     osname=darwin, osvers=8.0, archname=darwin-thread-multi-2level
     uname='darwin b28.apple.com 8.0 darwin kernel version 7.5.0: thu  
mar 3 18:48:46 pst 2005; root:xnuxnu-517.99.13.obj~1release_ppc power  
macintosh powerpc '
     config_args='-ds -e -Dprefix=/usr -Dccflags=-g  -pipe  - 
Dldflags=-Dman3ext=3pm -Duseithreads -Duseshrplib'
     hint=recommended, useposix=true, d_sigaction=define
     usethreads=define use5005threads=undef useithreads=define  
usemultiplicity=define
     useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
     use64bitint=undef use64bitall=undef uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-g -pipe -fno-common -DPERL_DARWIN -no-cpp- 
precomp -fno-strict-aliasing -I/usr/local/include',
     optimize='-Os',
     cppflags='-no-cpp-precomp -g -pipe -fno-common -DPERL_DARWIN -no- 
cpp-precomp -fno-strict-aliasing -I/usr/local/include'
     ccversion='', gccversion='3.3 20030304 (Apple Computer, Inc.  
build 1809)', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=8
     ivtype='long', ivsize=4, nvtype='double', nvsize=8,  
Off_t='off_t', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags ='-L/usr/ 
local/lib'
     libpth=/usr/local/lib /usr/lib
     libs=-ldbm -ldl -lm -lc
     perllibs=-ldl -lm -lc
     libc=/usr/lib/libc.dylib, so=dylib, useshrplib=true,  
libperl=libperl.dylib
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags='-bundle -undefined dynamic_lookup -L/ 
usr/local/lib'

Locally applied patches:
     23953 - fix for File::Path::rmtree CAN-2004-0452 security issue
     33990 - fix for setuid perl security issues


@INC for perl v5.8.6:
     /sw/lib/perl5
     /sw/lib/perl5/darwin
     /System/Library/Perl/5.8.6/darwin-thread-multi-2level
     /System/Library/Perl/5.8.6
     /Library/Perl/5.8.6/darwin-thread-multi-2level
     /Library/Perl/5.8.6
     /Library/Perl
     /Network/Library/Perl/5.8.6/darwin-thread-multi-2level
     /Network/Library/Perl/5.8.6
     /Network/Library/Perl
     /System/Library/Perl/Extras/5.8.6/darwin-thread-multi-2level
     /System/Library/Perl/Extras/5.8.6
     /Library/Perl/5.8.1/darwin-thread-multi-2level
     /Library/Perl/5.8.1
     .


Environment for perl v5.8.6:
     DYLD_LIBRARY_PATH (unset)
     HOME=/Users/juanfc
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=/sw/bin:/sw/sbin:/Library/Tcl/bin:/bin:/sbin:/usr/bin:/usr/ 
sbin:/Developer/Tools:/usr/X11R6/bin:/usr/local/teTeX/bin/powerpc- 
apple-darwin-current:/Users/juanfc/bin:/Users/juanfc/ubin:/Users/ 
juanfc/perls:/opt/local/bin:.:/usr/local/bin
     PERL5LIB=/sw/lib/perl5:/sw/lib/perl5/darwin
     PERL_BADLANG (unset)
     SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @rgarcia

On 3/17/06\, via RT ende \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> while \(1\) \{
>      \`/Users/home/bin/tclprocThatusesdevtty\.tcl \-a\`;

What happens if you use system\(\) instead of \`\` ? system doesn't
intercept the output of the command like \`\` does \-\- \`\` being intended
to capture output\, not to simply run an external program\.

>      \# \.\.\.\.\.
>      sleep 600;
> \}
>
> only the first time the Tcl program is runned\, the messages sent
> by the Tcl program are really sent to the /dev/tty\.  Then\, no more
> messages are seen in the Terminal\.  The perl script seems
> running ok all the time and the Tcl program also seems been called\.\.\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From shouldbedomo@mac.com

On 2006â??03â??17\, at 09&#8203;:40\, ende \(via RT\) wrote&#8203;:

> I send messages through /dev/tty with
>
>     exec echo >/dev/tty
>
> in a tcl script that is called from a Perl script\.
>
> The Perl script is on forever with a loop\, allowing me
> a more easy control of both the Tcl script and the frequency
> and actions I want the daemon do each 10 minutes\.

Can you make a short script \(or pair of scripts\) in pure Perl that  
show the problem? I ask because\, if the process really is a daemon\,  
it should not have a controlling terminal\, and so /dev/tty should not  
be openable at all \(not even once\)\. Here's an example on Mac OS X  
\(your platform \-\- basically BSD\-ish\)&#8203;:

$ perl \-MPOSIX=setsid \-lwe 'unless \(fork\) \{setsid; qx\(tty > /dev/ 
tty\); sleep 300\}'
sh&#8203;: line 1&#8203;: /dev/tty&#8203;: Device not configured

and the same on PPC Linux&#8203;:

$ perl \-MPOSIX=setsid \-lwe 'unless \(fork\) \{setsid; qx\(tty > /dev/ 
tty\); sleep 300\}'
sh&#8203;: /dev/tty&#8203;: No such device or address

\(In both cases\, we are seeing a complaint sent by the shell to  
standard error\, which is still open\, and which is not captured by qx\(\)\.\)
\-\- 
Dominic Dunlop

**Comment by p5pRT** [comment]
### From ende@telefonica.net



\#\!/usr/local/bin/perl
\#
\# it is the 5\.8\.8
\#
\# drgrun\.pl
\#
\# this is script that never ends \(is \_like\_ a daemon\)
\#
\#

while \(1\) \{
&nbsp;     \`/Users/me/bin/script\.tcl \-a\`;

&nbsp;     sleep 600;
\}

\# END

\# \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
\# Tcl script script\.tcl

eval \[list exec echo >/dev/tty  " \-\->> \\\[$cola\\\] \[string range  
\[urldecod \[file tail $que\]\] 0 $cosaEnActualL\]" \]


\# \#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

\# this script calls programs that could use stdout

\# The first time the script is called the echo > /dev/tty works but\,  
then
\# it gets dummy\.  If you run it in sh

\#\!/bin/bash

while \[ 1 \-eq 1 \]; do
&nbsp;     /Users/me/bin/script\.tcl \-a

&nbsp;     sleep 600
done

\# all goes well\, no problem\.  bash does not interfere with the output of
\# the Tcl script\.  I don't like to use sh because it uses an external
\# program\, sleep\.\.\.



El 17/03/2006\, a las 16&#8203;:05\, Dominic Dunlop via RT escribió&#8203;:

> On 2006â€“03â€“17\, at 09&#8203;:40\, ende \(via RT\) wrote&#8203;:
>
>> I send messages through /dev/tty with
>>
>>     exec echo >/dev/tty
>>
>> in a tcl script that is called from a Perl script\.
>>
>> The Perl script is on forever with a loop\, allowing me
>> a more easy control of both the Tcl script and the frequency
>> and actions I want the daemon do each 10 minutes\.
>
> Can you make a short script \(or pair of scripts\) in pure Perl that
> show the problem? I ask because\, if the process really is a daemon\,
> it should not have a controlling terminal\, and so /dev/tty should not
> be openable at all \(not even once\)\. Here's an example on Mac OS X
> \(your platform \-\- basically BSD\-ish\)&#8203;:
>
> $ perl \-MPOSIX=setsid \-lwe 'unless \(fork\) \{setsid; qx\(tty > /dev/
> tty\); sleep 300\}'
> sh&#8203;: line 1&#8203;: /dev/tty&#8203;: Device not configured
>
> and the same on PPC Linux&#8203;:
>
> $ perl \-MPOSIX=setsid \-lwe 'unless \(fork\) \{setsid; qx\(tty > /dev/
> tty\); sleep 300\}'
> sh&#8203;: /dev/tty&#8203;: No such device or address
>
> \(In both cases\, we are seeing a complaint sent by the shell to
> standard error\, which is still open\, and which is not captured by qx 
> \(\)\.\)
> \-\- 
> Dominic Dunlop
>
>

**Comment by p5pRT** [comment]
### From @dcollinsn

This ticket badly needs to be reviewed by a Darwin expert\. Is there something that we need to work around on this OS? \(Can you summarize it and update the ticket name?\)

\-\- 
Respectfully\,
Dan Collins

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

This ticket badly needs to be reviewed by a Darwin expert\. Is there something that we need to work around on this OS? \(Can you summarize it and update the ticket name?\)

\-\- 
Respectfully\,
Dan Collins

**Comment by p5pRT** [comment]
@dcollinsn - Status changed from 'open' to 'stalled'

**Comment by p5pRT** [comment]
### From @arc

Dan Collins via RT \<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:
> This ticket badly needs to be reviewed by a Darwin expert\. Is there something that we need to work around on this OS? \(Can you summarize it and update the ticket name?\)

I've tried and failed to reproduce this&#8203;:

$ uname \-sr
Darwin 13\.4\.0
$ perl \-wE 'say $^V'
v5\.24\.0
$ echo 'puts $tcl\_version' | tclsh
8\.6
$ cat outer\.pl
\#\! /usr/bin/env perl
while \(1\) \{
&nbsp;    \`\./inner\.tcl \-a\`;
&nbsp;    sleep 2;
\}
$ cat inner\.tcl
\#\! /usr/bin/env tclsh
exec echo foo >/dev/tty
$ chmod \+x outer\.pl inner\.tcl
$ \./outer\.pl
foo
foo
foo
^C

It seems to me that whatever bug existed when this ticket was filed a
decade ago has subsequently been fixed\, and I suggest this ticket can
therefore be closed now\. Alternatively\, if the original requestor can
submit complete\, self\-contained scripts that still reproduce the
problem \(the incomplete Tcl program in transaction 131257 merely fails
with a "no such variable" error before it runs anything at all\)\, we
can surely investigate further\.

\-\- 
Aaron Crane \*\* http&#8203;://aaroncrane\.co\.uk/

**Comment by p5pRT** [comment]
### From @dcollinsn

On Tue Jul 12 10&#8203;:18&#8203;:05 2016\, arc wrote&#8203;:
> Dan Collins via RT \<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:
> > This ticket badly needs to be reviewed by a Darwin expert\. Is there
> > something that we need to work around on this OS? \(Can you summarize
> > it and update the ticket name?\)
> 
> I've tried and failed to reproduce this&#8203;:
> 
> $ uname \-sr
> Darwin 13\.4\.0
> $ perl \-wE 'say $^V'
> v5\.24\.0
> $ echo 'puts $tcl\_version' | tclsh
> 8\.6
> $ cat outer\.pl
> \#\! /usr/bin/env perl
> while \(1\) \{
>     \`\./inner\.tcl \-a\`;
>     sleep 2;
> \}
> $ cat inner\.tcl
> \#\! /usr/bin/env tclsh
> exec echo foo >/dev/tty
> $ chmod \+x outer\.pl inner\.tcl
> $ \./outer\.pl
> foo
> foo
> foo
> ^C
> 
> It seems to me that whatever bug existed when this ticket was filed a
> decade ago has subsequently been fixed\, and I suggest this ticket can
> therefore be closed now\. Alternatively\, if the original requestor can
> submit complete\, self\-contained scripts that still reproduce the
> problem \(the incomplete Tcl program in transaction 131257 merely fails
> with a "no such variable" error before it runs anything at all\)\, we
> can surely investigate further\.

Thank you\, Aaron\.

Taking with the intent to close in 7 days unless there is anyone who can provide a test case that we can reproduce\.

\-\- 
Respectfully\,
Dan Collins

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'stalled' to 'open'

**Comment by toddr** [comment]
I think this case was not closed by mistake. Closing it now.
